authorization/v2/uniformResourceIdentifier.go

```diff
@@ -157,3 +158,49 @@ func marshallURIMap(uriMap map[string]string) string {
 	// Remove the extra '&' character introduced when marshaling the uriMap
 	return marshalledMap[:len(marshalledMap)-1]
 }
+
+// isURIMatch checks that the target URI is a subset of the source URI (the URI which the user can access)
+func isURIMatch(source UniformResourceIdentifier, target UniformResourceIdentifier) bool {
+	// Ensure the target path is a subset of the source path
+	if len(source.path) > len(target.path) {
+		return false
+	}
+
+	// Compare URI path
+	for i, sourceCharacter := range source.path {
+		if uint8(sourceCharacter) != target.path[i] {
+			return false
+		}
+	}
+
+	// Validate URI arguments
+	for key, sourceValue := range source.arguments {
+		if targetValue, ok := target.arguments[key]; ok {
+			match, err := regexp.Match(sourceValue, []byte(targetValue))
+			if !match || err != nil {
+				// Fail-soft, if the regex is invalid or the regex pattern match fails, the URIs don't match
+				return false
+			}
+		}
+	}
+
+	// Validate URI metadata
+	for key, sourceValue := range source.metadata {
+		targetValue, ok := target.metadata[key]
+		if !ok || sourceValue != targetValue {
+			return false
+		}
+	}
+
+	return true
+}
+
+// CompareURI validates that the source URI (the URI for which the user has permissions) matches
+// one of the URIs in the target set
+func (uri UniformResourceIdentifier) isSubsetOfAtLeastOne(targets []UniformResourceIdentifier) bool {
+	targetsMatch := false
+	for i := 0; !targetsMatch && i < len(targets); i++ {
+		targetsMatch = isURIMatch(uri, targets[i])
+	}
+	return targetsMatch
```

Comment:
I'd say something like this would be more readable:
```
for i := 0;  i < len(targets); i++ {
		if isURIMatch(uri, targets[i]) {
                                return true
                }
}
return false
```

Reply:
Yep, definitely